apps/cpp_rpc/rpc_tracker_client.h

```diff
@@ -0,0 +1,249 @@
+        if ((ready <= 0) || (!selecter.CheckRead(listen_sock.sockfd))) {
+          std::ostringstream ss;
+          ss << "[" << int(TrackerCode::kGetPendingMatchKeys) << "]";
+          tracker_sock_.SendBytes(ss.str());
+
+          // Receive status and validate
+          std::string pending_keys = tracker_sock_.RecvBytes();
+          old_keyset_.insert(*matchkey);
+
+          // if match key not in pending key set
+          // it means the key is acquired by a client but not used.
+          if (pending_keys.find(*matchkey) == std::string::npos) {
+              unmatch_period_count += 1;
+          } else {
+              unmatch_period_count = 0;
+          }
+          // regenerate match key if key is acquired but not used for a while
+          if (unmatch_period_count * ping_period > unmatch_timeout + ping_period) {
+            LOG(INFO) << "no incoming connections, regenerate key ...";
+
+            *matchkey = RandomKey(key_ + ":", old_keyset_);
+
+            std::ostringstream ss;
+            ss << "[" << static_cast<int>(TrackerCode::kPut) << ", \"" << key_ << "\", ["
+               << port << ", \"" << *matchkey << "\"], " << custom_addr_ << "]";
+            tracker_sock_.SendBytes(ss.str());
+
+            std::string remote_status = tracker_sock_.RecvBytes();
+            CHECK_EQ(std::stoi(remote_status), static_cast<int>(TrackerCode::kSuccess));
+            unmatch_period_count = 0;
+          }
+          continue;
+        }
+      }
+      break;
+    }
+  }
+
+ private:
+  /*!
+   * \brief Connect to a RPC address with retry.
+            This function is only reliable to short period of server restart.
+   * \param timeout Timeout during retry
+   * \param retry_period Number of seconds before we retry again.
+   * \return TCPSocket The socket information if connect is success.
+   */
+  common::TCPSocket ConnectWithRetry(int timeout = 60, int retry_period = 5) {
+    auto tbegin = std::chrono::system_clock::now();
+    while (1) {
+      common::SockAddr addr(tracker_addr_);
+      common::TCPSocket sock;
+      sock.Create();
+      LOG(INFO) << "Tracker connecting to " << addr.AsString();
+      if (sock.Connect(addr)) {
+        return sock;
+      }
+
+      auto period = (std::chrono::duration_cast<std::chrono::seconds>(
+                  std::chrono::system_clock::now() - tbegin)).count();
+      CHECK(period < timeout) << "Failed to connect to server" << addr.AsString();
+      LOG(WARNING) << "Cannot connect to tracker " << addr.AsString()
+                   << " retry in " << retry_period << " seconds.";
+      std::this_thread::sleep_for(std::chrono::seconds(retry_period));
+    }
+  }
+  /*!
+  * \brief Random Generate a random number between 0 and 1.
+  * \return random float value.
+  */
+  float Random() {
+    std::random_device rd;  // Will be used to obtain a seed for the random number engine
+    std::mt19937 gen(rd());  // Standard mersenne_twister_engine seeded with rd()
```

Comment:
gen and distribution should be member of the class, as random number generator needs to be stateful